### PR TITLE
Docker compose templating

### PR DESCRIPTION
**Pull request by ghost**
instrument based on yq and bash for quick docker compose formatting

### CODE REVIEW

File: Scripts/01_bash/bash_templating/readme.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+## Commands
+* -cp=value - name of command property for replacement
+* -cv=value - value of command property for replacement
+* -cpr=value - command property for removing
+* -i=value - new image 
+
+## Usage example
+```bash
+chmod u+x templating.sh
```
Comment (nazar-pc) [inline Scripts/01_bash/bash_templating/readme.md]: It is already tracked by Git, no need to run it if you commit executable file in the first place

File: Scripts/01_bash/bash_templating/target-docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      "--reserved-nodes", "/dns/bootstrap-9.gemini-1b.subspace.network/tcp/30333/p2p/12D3KooWNGxWQ4sajzW1akPRZxjYM5TszRtsCnEiLhpsGrsHrFC6",
+      "--reserved-only",
+    ]
+    healthcheck:
+      timeout: 5s
+# If node setup takes longer then expected, you want to increase `interval` and `retries` number.
+      interval: 30s
+      retries: 5
+volumes:
+  node-data:
```
Comment (nazar-pc) [inline Scripts/01_bash/bash_templating/target-docker-compose.yml]: Please configure your IDE to always add new lines to the end of text files automatically

**Review by nazar-pc — COMMENTED** [review]
This seems too complex to maintain and is single-purpose, I'm really not sure we want it.

Something like this would have been much easier and would work for any files, not just YAML:
```
version: "3.7"
services:
  node:
    image: IMAGE_NAME
    volumes:
      - node-data:/var/subspace:rw
    ports:
      - "0.0.0.0:30333:30333"
    restart: unless-stopped
    command: [
      "--chain", "gemini-1",
      "--base-path", "/var/subspace",
      "--execution", "wasm",
      "--pruning", "archive",
      "--listen-addr", "/ip4/0.0.0.0/tcp/30333",
      "--in-peers", IN_PEERS,
      "--in-peers-light", "2000",
      "--out-peers", "1000",
      "--reserved-nodes", "/dns/bootstrap-9.gemini-1b.subspace.network/tcp/30333/p2p/12D3KooWNGxWQ4sajzW1akPRZxjYM5TszRtsCnEiLhpsGrsHrFC6",
      "--reserved-only",
    ]
    healthcheck:
      timeout: 5s
# If node setup takes longer then expected, you want to increase `interval` and `retries` number.
      interval: 30s
      retries: 5
volumes:
  node-data:
```
```
./templating.sh \
    --replace IMAGE_NAME=ghcr.io/subspace/node:gemini-1b-2022-jun-13 \
    --replace IN_PEERS=100 \
    /path/to/docker-compose.yml
```

**Comment by ghost** [comment]
> This seems too complex to maintain and is single-purpose, I'm really not sure we want it.
> 
> Something like this would have been much easier and would work for any files, not just YAML:
> 
> ```
> version: "3.7"
> services:
>   node:
>     image: IMAGE_NAME
>     volumes:
>       - node-data:/var/subspace:rw
>     ports:
>       - "0.0.0.0:30333:30333"
>     restart: unless-stopped
>     command: [
>       "--chain", "gemini-1",
>       "--base-path", "/var/subspace",
>       "--execution", "wasm",
>       "--pruning", "archive",
>       "--listen-addr", "/ip4/0.0.0.0/tcp/30333",
>       "--in-peers", IN_PEERS,
>       "--in-peers-light", "2000",
>       "--out-peers", "1000",
>       "--reserved-nodes", "/dns/bootstrap-9.gemini-1b.subspace.network/tcp/30333/p2p/12D3KooWNGxWQ4sajzW1akPRZxjYM5TszRtsCnEiLhpsGrsHrFC6",
>       "--reserved-only",
>     ]
>     healthcheck:
>       timeout: 5s
> # If node setup takes longer then expected, you want to increase `interval` and `retries` number.
>       interval: 30s
>       retries: 5
> volumes:
>   node-data:
> ```
> 
>  ```
> ./templating.sh \
>     --replace IMAGE_NAME=ghcr.io/subspace/node:gemini-1b-2022-jun-13 \
>     --replace IN_PEERS=100 \
>     /path/to/docker-compose.yml
> ```

hmm, updating parameters yes, just variable will do. But there's no way to remove lines safely and in programatic way

**Comment by nazar-pc** [comment]
> hmm, updating parameters yes, just variable will do. But there's no way to remove lines safely and in programatic way

Do we need it though? Honestly I don't expect us to change many parameters anyway. It is easy to edit the `docker-compose.yml` directly and we would only need to change parameters for special cases where there are multiple machines and some specific value is different for them. For that we need templating. Am I missing something?

**Comment by ghost** [comment]
> > hmm, updating parameters yes, just variable will do. But there's no way to remove lines safely and in programatic way
> 
> Do we need it though? Honestly I don't expect us to change many parameters anyway. It is easy to edit the `docker-compose.yml` directly and we would only need to change parameters for special cases where there are multiple machines and some specific value is different for them. For that we need templating. Am I missing something?

Today our yaml is pretty much uniform in terms of parameters and arguments but in future there could different DCs and on bare metal and with different commands/directives. Script could updated to read diffs from git. It was intentionally written in bash for easy pipeline/ansible integration. Also in future it could be used to update consolidated state of DCs of 100s machines. So I think it would be a good trial to see if it fits the overall workflow

**Comment by nazar-pc** [comment]
Hm... if we have different DCs where configs are significantly different, why wouldn't we have different templates instead of one template for everything and heavy use of `templating.sh`? I just don't see how it is more convenient.

Ideally we wouldn't need anything at all and used a capabilities native for deployment tool, like native templating in helm charts.

**Comment by ghost** [comment]
Agree but that would bring dependency on service provider as on-premise k8s adds another layer of complexity. Each new bare-metal provider would require master and stable network. Each new release introduces challenges and require migration analysis.
In most cases blockchain companies settle for one or another provider and they are not actually decentralized. AWS + GCP doesn’t count for decentralization. `docker-compose` Is really easy to run. Templating script could be added to cron to update image or even to container itself, so it will be updated automatically. Also compose itself could be expanded further to accommodate synthetic tests/healthchecks and Prometheus exporters/pushgateway 
So now I see it as only option to run nodes on 10s bare metal providers and 1000s machines(availability can achieved by number of machines + LB of ws and not by quality of provider) Correct me If I am wrong plz

**Comment by nazar-pc** [comment]
I did not say we need to have k8s on bare metals, it was just an example.

I'm sure there are tools that support templating. For instance I used Chef in the past, it supports templating, can be used for docker-compose.yml too with reasonably good syntax actually. And there is a cookbook for Docker there as well, it supports defining declaratively what container should look like and if it doesn't match that in practice - will be recreated.

Far from help and k8s, but it works reasonably well even without docker-compose.yml at all.

**Comment by ghost** [comment]
this one correct https://supermarket.chef.io/cookbooks/docker ?

**Comment by nazar-pc** [comment]
I think so. Again, Chef is something I used, though not particularly enjoyed, just another example. Maybe there is something better, I'm not up to date with the tooling available.

**Comment by ghost** [comment]
https://kruschecompany.com/create-cloud-with-docker-rancher/

Rancher could be the tool we need. Meanwhile there's no good tools that can handle directive manipulation (add/ remove) in docker compose

**Comment by nazar-pc** [comment]
> Meanwhile there's no good tools that can handle directive manipulation (add/ remove) in docker compose

But we only need string replacement, not add/remove kind of manipulation here. A few lines of Python will get us there: read the file, read CLI arguments, replace string, write to stdout.

We can merge this as is, but only if the plan is for it to live a very short life because I neither see the need for it nor I see it as a maintainable tool unfortunately.

**Comment by ghost** [comment]
Yes, lets merge please. It was meant to be temporary :). At later stages Rancher or similar tool will take over.

**Review by nazar-pc — APPROVED** [review]
Approving just so we can merge at least something.

Please squash on merge, there are too many commits here.

**Comment by nazar-pc** [comment]
Wasn't squashed on merge :disappointed:

**Pull request merged**